allennlp/data/semparse/type_declarations/type_declaration.py

```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
             else:
                 raise RuntimeError("Unknown prefix: %s. Did you forget to pass it to the constructor?" % prefix)
         return super(DynamicTypeLogicParser, self).make_VariableExpression(name)
+
+
+def _substitute_any_type(_type: Type, basic_types: Set[BasicType]) -> Set[Type]:
+    """
+    Takes a type, and a set of basic types and substitutes all instances of ANY_TYPE with all possible basic
+    types, and returns a set with all possible combinations.
+    Note that this substitution is unconstrained. That is, If you have a type with placeholders,
+    <#1,#1> for example, this may substitute the placeholders with different basic types. In that case, you'd
+    want to use ``_substitute_placeholder_type`` instead.
+    """
+    if _type == ANY_TYPE:
+        return basic_types
+    if isinstance(_type, (BasicType, PlaceholderType)):
+        return set([_type])
+    substitutions = set()
+    for first_type in _substitute_any_type(_type.first, basic_types):
+        for second_type in _substitute_any_type(_type.second, basic_types):
+            substitutions.add(ComplexType(first_type, second_type))
+    return substitutions
+
+
+def _substitute_placeholder_type(_type: Type, basic_type: BasicType) -> Type:
+    """
+    Takes a type with placeholders, and a basic type and substitutes all occurrences of the placeholder with
+    that type.
+    """
+    # TODO (pradeep): This assumes there's just one placeholder in the type. So this doesn't work with
+    # ``reverse`` yet, which has two placeholders.
+    if _type == ANY_TYPE:
+        return basic_type
+    if isinstance(_type, BasicType):
+        return _type
+    return ComplexType(_substitute_placeholder_type(_type.first, basic_type),
+                       _substitute_placeholder_type(_type.second, basic_type))
+
+
+def _make_production_string(source: Type, target: Union[List[Type], Type]) -> str:
+    return "%s -> %s" % (str(source), str(target))
+
+
+def _get_complex_type_productions(complex_type: ComplexType) -> List[Tuple[str, str]]:
+    """
+    Takes a complex type without any placeholders and returns all productions that lead to it, starting
+    from the most basic return type. For example, if the complex is `<a,<<b,c>,d>>`, this gives the
+    following tuples
+    ('<<b,c>,d>', '<<b,c>,d> -> [<a,<<b,c>,d>>, a]')
+    ('d', 'd -> [<<b,c>,d>, <b,c>]')
```

Comment:
Why doesn't this also give `('c', 'c -> [<b,c>, b]')`?  Or does it?  It looks to me like the code will produce this; maybe you just missed it in the docstring?

Reply:
It didn't do that earlier. I expanded only the branches that lead to the complex type, but I could have expanded the others as well. Changed this now. Thanks for pointing this out!